### PR TITLE
Helpers functions and weighted datastream sampling

### PR DESCRIPTION
**Pull request by samedii**
- Weighted sampling from datastreams
- State dicts for datastreams/samplers
- Worker init function for settings seeds correctly
- Train and eval decorators handle output automatically and moves tensors to cpu (and detaches)
- Decorator for updating cpu clone of model
- Soft binary cross entropy

### CODE REVIEW

**Review by FelixAbrahamsson — APPROVED** [review]
Great stuff, we should publish 0.7.2 after merging this

File: workflow/functional/structure_map.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+
+
+def structure_map(fn, x):
```
Comment (Jim-Holmstroem) [inline workflow/functional/structure_map.py:3]: usually called `fmap` short for `functor map` https://wiki.haskell.org/Functor

File: workflow/ignite/decorators/evaluate.py [hunk 2]
```diff
@@ -1,21 +1,32 @@
...
+import torch
 from functools import wraps
 
+from workflow.functional import structure_map
 from workflow.torch import model_device
 from workflow.ignite.decorators.to_device import to_device
 
 
+def detach(x):
+    if type(x) is torch.Tensor:
```
Comment (Jim-Holmstroem) [inline workflow/ignite/decorators/evaluate.py]: if something is a subtype it should also detach
https://docs.python.org/3/library/functions.html#issubclass

File: workflow/torch/datastream.py [hunk 3]
```diff
@@ -1,84 +1,343 @@
...
+        self.weights[:] = state_dict['weights']
+
+
+class MergeSampler(torch.utils.data.Sampler):
+    def __init__(self, samplers, datasets, ns):
+        self.samplers = samplers
+        self.datasets = datasets
+        self.ns = ns
+        self.from_mapping = Dataset.create_from_concat_mapping(datasets)
+        self.merged_samplers = MergeSampler.merge_samplers(
```
Comment (Jim-Holmstroem) [inline workflow/torch/datastream.py:53]: confused it with `def merge_samplers` took a bit to realize

File: workflow/torch/datastream.py [hunk 4]
```diff
@@ -1,84 +1,343 @@
...
+
+    def __len__(self):
+        return self.length
+
+    def __iter__(self):
+        return iter(self.merged_samplers)
+
+    @staticmethod
+    def merged_samplers_length(samplers):
+        # TODO: this is probably incorrect?
```
Comment (Jim-Holmstroem) [inline workflow/torch/datastream.py]: when should we deal with this?

File: workflow/ignite/decorators/evaluate.py (hunk 2 above)
Comment (samedii) [reply workflow/ignite/decorators/evaluate.py]: Yes maybe, I think we should use `isinstance` though. I can change that

File: workflow/torch/datastream.py [hunk 5]
```diff
@@ -1,84 +1,343 @@
...
+
+    @staticmethod
+    def merge_samplers(samplers, datasets, ns):
+        to_mapping = Dataset.create_to_concat_mapping(datasets)
+
+        def batch(iterable, n):
+            while True:
+                yield [next(iterable) for _ in range(n)]
+
+        index_batch = zip(*[
```
Comment (Jim-Holmstroem) [inline workflow/torch/datastream.py:79]: the unpacking `*` is not lazy, but it will probably never be an issue

File: workflow/torch/datastream.py [hunk 6]
```diff
@@ -171,3 +435,93 @@ def test_datastream_zip():
...
+    for attempt in range(10):
+        print('attempt:', attempt)
+        datastream = Datastream.merge([
+            (ZippedMergedDatastream(), 1),
+            (ZippedMergedDatastream(), 5),
+        ])
+
+        it = iter(datastream.data_loader(batch_size=16, n_batches_per_epoch=10))
+        for _ in range(10):
+            print(next(it))
```
Comment (Jim-Holmstroem) [inline workflow/torch/datastream.py:465]: tests should be silent unless they fail :)

**Review by Jim-Holmstroem — COMMENTED** [review]
mostly minor comments, great work on this :D

File: workflow/ignite/decorators/evaluate.py (hunk 2 above)
Comment (Jim-Holmstroem) [reply workflow/ignite/decorators/evaluate.py]: sorry, mixed it up, still a bit rusty in the weirder parts of python :P
yes isinstance is the most pythonic way I think

also I think there where a few more instances of is torch.Tensor :)

File: workflow/torch/datastream.py (hunk 5 above)
Comment (FelixAbrahamsson) [reply workflow/torch/datastream.py:79]: The length of that list is bounded by the nr of samplers so it's always fine.

File: workflow/torch/datastream.py (hunk 4 above)
Comment (FelixAbrahamsson) [reply workflow/torch/datastream.py]: The behavior of this function has been discussed a lot, there are arguments for all kinds of variants but we settled on this because it has the most practical convenience. We can remove the `TODO` comment I think.

File: workflow/functional/structure_map.py (hunk 1 above)
Comment (samedii) [reply workflow/functional/structure_map.py:3]: `fmap` seems to work on both key and value though

File: workflow/ignite/decorators/evaluate.py (hunk 2 above)
Comment (samedii) [reply workflow/ignite/decorators/evaluate.py]: The reason that I didn't use isinstance is that I'm not sure that we actually want to allow subclasses everywhere but I might be wrong

For example in `structure_map` we do not want to allow subclasses

File: workflow/functional/structure_map.py (hunk 1 above)
Comment (Jim-Holmstroem) [reply workflow/functional/structure_map.py:3]: no it applies over the inner type

File: workflow/torch/datastream.py (hunk 6 above)
Comment (samedii) [reply workflow/torch/datastream.py:465]: pytest only prints if there is an error so I thought this was intended. Does that change your mind?

File: workflow/functional/structure_map.py (hunk 1 above)
Comment (Jim-Holmstroem) [reply workflow/functional/structure_map.py:3]: Look at the hashmap for example:

https://hackage.haskell.org/package/hashmap-1.3.3/docs/Data-HashMap.html

only `Map[K, .]` (i.e. `Map k` in the docs) is a functor, which is the "natural" choice

you can of course construct a reverse map and have the keys as the natural thing to map over :P

Comment (Jim-Holmstroem) [reply workflow/functional/structure_map.py:3]: it's perhaps a bit silly to refer to Haskell docs, but if they got something right its the way they type stuff

Comment (samedii) [reply workflow/functional/structure_map.py:3]: I don't know haskell but I interpreted this as f on both the key and value?
`fmap :: (a -> b) -> f a -> f b`

Comment (Jim-Holmstroem) [reply workflow/functional/structure_map.py:3]: perhaps its better to have something without the connotation that fmap brings :)

File: workflow/torch/datastream.py (hunk 6 above)
Comment (Jim-Holmstroem) [reply workflow/torch/datastream.py:465]: a thats right, it actually traps the output and only prints everything once it fails?
in that case and given the nature of what this is trying to test it's okay :)

Comment (Jim-Holmstroem) [reply workflow/torch/datastream.py:465]: just to add, I thought the prints was more of something you used when you constructed the test and then just let it be left

File: workflow/torch/datastream.py (hunk 3 above)
Comment (samedii) [reply workflow/torch/datastream.py:53]: We can have another look at this after splitting to a new library

File: workflow/ignite/decorators/evaluate.py (hunk 2 above)
Comment (Jim-Holmstroem) [reply workflow/ignite/decorators/evaluate.py]: subclasses of Tensor should still be a Tensor in all senses of the type, otherwise they have implemented it badly

as for inside `structure_map` the issue is that you want to construct the same type as you deconstruct
so yes, with this implementation you would want to keep it as == equals to maintain the property that the structure is preserved

it would be possible to use the dynamic nature of python and have
```
if isinstance(d, dict):
    return type(d)(map(lambda kv: (kv[0], function(kv[1])), d.items()))
```

Comment (Jim-Holmstroem) [reply workflow/ignite/decorators/evaluate.py]: this would support `OrderedDict` for example and other things that are sub of dict (as long as they have .items() and are able to construct from [(k, v)]

Comment (samedii) [reply workflow/ignite/decorators/evaluate.py]: Yes I considered what you're saying but I didn't want to make it too complicated. Maybe we should reevaluate that but I think it's good enough in the current condition and we can make a PR to change that if we want to

Comment (Jim-Holmstroem) [reply workflow/ignite/decorators/evaluate.py]: yes I think so to, as stated "in the current implementation"
I wasn't requesting a change for what structure_map supports
it will be backwards compatible if we add support for subtypes of the types listed later :)

**Review by Jim-Holmstroem — APPROVED** [review]
:dolphin:
